src/wallet/walletdb.cpp

```diff
@@ -405,6 +407,62 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
             ssValue >> keyMeta;
             wss.nKeyMeta++;
             pwallet->GetOrCreateLegacyScriptPubKeyMan()->LoadKeyMetadata(vchPubKey.GetID(), keyMeta);
+
+            // Extract some CHDChain info from this metadata if it has any
+            if (keyMeta.nVersion >= CKeyMetadata::VERSION_WITH_HDDATA && !keyMeta.hd_seed_id.IsNull() && keyMeta.hdKeypath.size() > 0) {
+                // Get the path from the key origin or from the path string
+                // Not applicable when path is "s" as that indicates a seed
+                bool internal = false;
+                uint32_t index = 0;
+                if (keyMeta.hdKeypath != "s") {
+                    std::vector<uint32_t> path;
+                    if (keyMeta.has_key_origin) {
+                        // We have a key origin, so pull it from it's path vector
+                        path = keyMeta.key_origin.path;
+                    } else {
+                        // No key origin, have to parse the string
+                        if (!ParseHDKeypath(keyMeta.hdKeypath, path)) {
+                            strErr = "Error reading wallet database: keymeta with invalid HD keypath";
+                            return false;
+                        }
+                    }
+
+                    // Extract the index and internal from the path
+                    // Path string is m/0'/k'/i'
+                    // Path vector is [0', k', i'] (but as ints OR'd with the hardened bit
+                    // k == 0 for external, 1 for internal. i is the index
+                    if (path.size() != 3) {
+                        strErr = "Error reading wallet database: keymeta found with unexpected path";
+                        return false;
+                    }
+                    if (path[0] != (0 & 0x80000000)) {
+                        strErr = "Unexpected path index for the element at index 0";
+                    }
+                    if (path[1] != (0 & 0x80000000) && path[1] != (1 & 0x80000000)) {
```

Comment:
In commit "Determine inactive HD seeds from key metadata and track them in LegacyScriptPubKeyMan" (daa9e347deceedb354aa0c73e34e34d50fafb5d5)

Doesn't & need to be | here (and above) (and below)? I'm actually not sure how the test seems to pass with this

Reply:
> I'm actually not sure how the test seems to pass with this

Oh, there actually are "Unexpected path index" errors in the wallet_hd.py debug logs. I guess these errors don't prevent the wallet from being loaded like I initially thought https://github.com/bitcoin/bitcoin/pull/17681#discussion_r421539430

Reply:
No, you are correct, they should be `|`. The lack of failure is because I forgot to `return false`.